Outline how to use the new -generateTrace switch (#262)
microsoft/TypeScript-wiki: Performance-Tracing.md

# Performance Tracing

TypeScript 4.1 introduced a `generateTrace` flag to make it easier to identify the parts of a program that are taking the most time to compile.

## Warnings

This feature is still experimental and will likely change significantly in TS 4.2.

> In particular, though it would be easy and helpful to post-process the trace files, you should assume that your post-processing tools will be broken by TS 4.2.

Reviewer: ```suggestion
In particular, though it would be easy and helpful to post-process the trace files, you should assume that any post-processing tools that may be built will be broken by TS 4.2.
```

Author: "your" -> "any such"

Trace files are basically useless without the underlying code, so there's little reason to share one with anyone who can't access your code.
If you do share them, please consider zipping them - they compress very well.

## Getting Started

At present, the output is not very user friendly, but here are some steps you can take to identify the parts of your code that are slowing the compiler down.
The goal of this process is to be able to extract a reduced repro for which you can file an [issue](https://github.com/microsoft/TypeScript/issues).

1.  This document is specific to TS 4.1 RC and TS 4.1, so you'll need to install one of those.

>     The easiest way is to `npm install typescript@rc` ( or `typescript@4.1`, once it is released).

Reviewer: ```suggestion
    The easiest way is to `npm install typescript@rc` (or `typescript@4.1`, once it is released).
```

>     There's no need to check in a change to your repo - you can just use it locally while you do your tracing.

Reviewer: ```suggestion
    Tracing with TypeScript 4.1 can be done once locally, so any changes to your `node_modules` and `package.json` can be discarded afterwards.
```

Author: I like it.

2.  Rebuild your project with the newly installed version of Typescript.
    Odds are you're going to get some new compiler errors.

>     Ideally, you want to fix these new errors correctly, so that the code you trace will be realistic.

Reviewer 2: "correctly" is redundant

Author: Technically, yes, but I wanted to somehow express that I didn't mean merely "mitigate".

Author: Reworded.

>     Ideally, you want to fix these new errors correctly, so that the code you trace will be realistic.

Reviewer: ```suggestion
    Ideally you'll need to fix these new errors so that the code you trace will be realistic.
```

Author: "Ideally, you'll want to address these new errors"

    If that is not feasible, prefer `@ts-ignore` to `any`, because it will cause less short-circuiting.

> 3.  Once your project builds without errors, compile your project with `tsc` and `-generateTrace some_directory`.

Reviewer 2: `-generateTrace` -> `--generateTrace` (I see a single dash)

Also in `-incremental` below

Reviewer: Either works technically, but I prefer 2 dashes just to give people a clue.

Author: We accept both and I believe the help text suggests just one.

Author: The help text uses two, so I'll match that.

Reviewer 2: I didn't even know that it accepts a single dash.  Is this some node convention?  (Usually `-ab` is equivalent to `-a -b`, and therefore long options require `--`)

    You can use `.` as the directory, but it may get cluttered, so a subdirectory is usually preferable.
    In this early version of the feature, you have to use `tsc` specifically - building through a bundler that invokes TypeScipt via the API will not work.

>     For best results, make sure this is not an incremental build (i.e. pass `-f` in build mode or `-incremental false` for regular compilation).

Reviewer: It might be a good idea to provide an example of just running the command fully, and ensuring that users know they're running the correct tsc.

Unless they are using something like Volta, you might want to just say to always use a global `tsc`?

Author: Can I just give `npx` and let yarn users figure it out or do I need both?

    1.  Use the little triangle to expand the view of the "Main" thread.
        For convenience, we ignore all other threads.
    2.  This little toolbar is floating - you'll probably want to move it into the blank space under the thread name (i.e. into the left sidebar).
        Each button indicates a cursor mode.
    3.  This mode is for selection.
        Selecting a single element is possible in most modes, but box selection is only possible in this one.\
        *Tip:* box selection is a great way to get a summary of a region.
    4.  This mode is for panning.
        Click-drag to scroll to other parts of the trace..
    5.  This mode is for zooming.
        Click-drag up or right to zoom in and down or left to zoom out.

7.  If you expand the Main thread and select one of the boxes, you'll see something like this, with time on the horizontal axis and (partial) call stacks growing down from the top:\
    ![Expanded view of edge://tracing](https://raw.githubusercontent.com/wiki/Microsoft/TypeScript/images/tracingExpanded.png)
    1.  You can easily differentiate the four major phases of compilation.

>         This is program construction, which includes parsing and modules resolution.

Reviewer 2: "module resolution"?

Author: Oops.

>         This is program construction, which includes parsing and modules resolution.

Reviewer: ```suggestion
        This is program construction, which includes parsing and module resolution.
```

Author: Fixed.

        Usually, you want to look for the largest `findSourceFile` boxes and follow them down to see which dependency chain is taking the most time.

>         Usually, it will be a library outside your own code.

Reviewer 2: Double "Usually" looks weird.

Author: Changed the second one to "Generally".

        Be sure to include this library in any repro you construct.
    2.  This is the binding phase.
        It is rarely an issue, so we'll ignore it for now.
    3.  This is the (type) checking phase, where most of the TypeScript-specific work happens.
        *You generally want to start here.*

>         We'll zoom in in a later step.

Reviewer 2: Maybe "We'll focus on it further below"?  (To avoid the "in in")

    4.  This is the emit phase, which we'll ignore for now.
    5.  This is the selected box.
    6.  These are the built-in metadata about the box.
        1.  `Start` is the timestamp of the start of the operation - usually uninteresting.
        2.  `Wall Duration` this is what would be called "Total Time" in many profiling tools.

>         3.  `Self Time` is the amount of time that is not accounted for by child boxes (i.e. directly below this one).

Reviewer 2: "Time spent in this box rather than in one of its descendants"?

    7.  This metadata is specific to the kind of box you're looking at.

>         *This is the most important information* - it will contain paths and things.

Reviewer 2: "and things" :)

>     *Note:* All times reflect the overhead of tracing, which is not spread uniformly through the execution.

Reviewer 2: Drop "All", and maybe mention that it will hopefully be reduced to not cause visible noise?

Author: I like the "all" but I've added a note about future improvements.

        1.  `checkSourceFile` gives the time spent checking a particular file.
            Look in the metadata for the file path.
        2.  `checkExpression` gives the time spent checking a particular expression.
            It will *usually, but not always* be in the file indicated by the containing `checkSourceFile`.
            As of now, there is no way to tell without looking at the code and no way to recover the path if it's not the one from `checkSourceFile`.
            We're thinking about how to handle it.
        3.  `checkVariableDeclaration` gives the time spent checking a particular variable declaration.
            Over the course of several investigations, we noticed that these were showing up often enough to merit their own boxes.
            The metadata for these boxes is not particularly convenient to consume.
            1.  `kind` identifies a member of the [SyntaxKind](https://github.com/microsoft/TypeScript/blob/9c60d5a4d33e4ed846cad2136da9efcd48adf847/src/compiler/types.ts#L21) enum.
            2.  `pos` is the zero-indexed character offset from the beginning of the file to the beginning of the range being checked.

>                 Note that position will be *before* any trivia (whitespace, comments) before the relevant code fragment, so it will frequently appears to be immediately after the *preceding* statement.

Reviewer 2: "will frequently appear" (drop the "s" from "appears")

            3.  `end` is the zero-indexed character offset from the beginning of the file to the end of the range being checked.
    2.  In this particular screenshot, we see that a single variable declaration accounts for a large percentage of the check time of this entire file.
        However, the declaration itself is not interesting because its child is an expression that is nearly as wide (i.e. the declaration has low "Self Time").
        This expression has four interesting sub-expressions, each of which has many, many children.
        So identifying those four ranges in the file would be the place to start when reducing a repro.

> 9.  Once you have identified the problematic code, review [[Performance]] for suggestions relevant to your scenario.

Author: This link isn't working in the preview and I don't know if that's a preview issue or if the syntax is wrong.

Author: It works on the real page.

Here's a slow section with no `checkExpression` boxes in the (partial) stack.
Instead, it's gone straight into type checking.
To work out which code fragment triggered this work and what it's doing, you'll need to open `types.json`.
From the metadata, we can see that the types of interest have IDs `20441` and `20440`.
Here are the corresponding entries from `types.json`.
(At the time of writing, the type with ID `N` is on line `N`.)

```
{"id":20440,"symbolName":"NamedExoticComponent","recursionId":30,"instantiatedType":146,"typeArguments":[20437],"firstDeclaration":{"path":"PROJECT_ROOT/node_modules/@types/react/index.d.ts","start":{"line":359,"character":6},"end":{"line":363,"character":6}},"flags":["524288"]},
{"id":20441,"symbolName":"NamedExoticComponent","recursionId":30,"instantiatedType":146,"typeArguments":[20434],"firstDeclaration":{"path":"PROJECT_ROOT/node_modules/@types/react/index.d.ts","start":{"line":359,"character":6},"end":{"line":363,"character":6}},"flags":["524288"]},
```

This data gives us some sense of what they look like and a location where we can review the declaration.

> In this case, we're comparing two different instantiations of [NamedExoticComponent&lt;P&gt;](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/599ef3c6e2142d87b6b4bcf030fb798504c3a468/types/react/index.d.ts#L361) - one instantiated with type `20434` and the other with `20437`.

Reviewer 2: Shouldn't you use `<` and `>` in a markdown link label?

Author: At least in VS Code, that inserted a paragraph break.

Author: I'll review the file when it's up and fix it if it has issues (mostly, links and images that can't be verified here).

Reviewer 2: Ah, I forgot that it allows arbitrary HTML.  *But* what's really missing there are backticks, so ``` [`NamedExoticComponent<P>`](...) ``` should work and look better...

Author: Switched to using backticks.

These types are also described in `types.json`, so keep looking up more types recursively until you understand what's going on.

> From the file path in `checkSourceFile` and the types we've just pulled out of `types.json`, we can probably guess approximately which code fragment was involved (knowing in advance that it's neither an expression, nor a variable declaration, which would have gotten a box).

Reviewer 2: "guess approximately" -> either "guess" or "approximate"?

Author: I think there's a difference between an exactly guess and an approximate guess.  Guessing which statement is the problem is different from narrowing it down to one of three candidates or a function body.

Look hard enough and you'll be able to create a toy program that reproduces the performance issue you're seeing in your real code.
We'd love to hear about it but, if you still can't share, the next step is to play around with alternate declarations and `any` type assertions until you've sped up the toy program.
Then you can apply the same fix or mitigation to your real code.

> *Aside:* the [edge://tracing](edge://tracing) tracing controls are rather clumsy.

Reviewer 2: Maybe mention in the above something like "see the bottom for using the usual devtools profiler"?

Also, maybe add that we'll try to make it work as the main tool used for visualization, if possible?

Author: I was de-emphasizing it because I thought you wanted to use a different tool.  If you've used it for an investigation and feel that it's up to the task, I don't mind moving this up.
